From 36c535907817f6fa92668fcab69e72077aff2bd7 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 28 Aug 2014 14:23:19 -0400
Subject: [cpufreq] intel_pstate: Add support for Baytrail turbo P states

Message-id: <1409235879-1283-106-git-send-email-prarit@redhat.com>
Patchwork-id: 89479
O-Subject: [RHEL7.1 PATCH BZ 1134639 105/185] intel_pstate: Add support for Baytrail turbo P states
Bugzilla: 1134369
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1134639

commit 61d8d2abc15e9232c3914c55502b73e559366583
Author: Dirk Brandewie <dirk.j.brandewie@intel.com>
Date:   Wed Feb 12 10:01:07 2014 -0800

    intel_pstate: Add support for Baytrail turbo P states

    A documentation update exposed the existance of the turbo ratio
    register. Update baytrail support to use the turbo range.

    Signed-off-by: Dirk Brandewie <dirk.j.brandewie@intel.com>
    Cc: 3.13+ <stable@vger.kernel.org> # 3.13+
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/cpufreq/intel_pstate.c b/drivers/cpufreq/intel_pstate.c
index 61c26c7..95e8968 100644
--- a/drivers/cpufreq/intel_pstate.c
+++ b/drivers/cpufreq/intel_pstate.c
@@ -34,8 +34,10 @@
 
 #define SAMPLE_COUNT  3
 
-#define BYT_RATIOS 0x66a
-#define BYT_VIDS        0x66b
+#define BYT_RATIOS  0x66a
+#define BYT_VIDS  0x66b
+#define BYT_TURBO_RATIOS 0x66c
+
 
 #define FRAC_BITS 8
 #define int_tofp(X) ((int64_t)(X) << FRAC_BITS)
@@ -367,6 +369,13 @@ static int byt_get_max_pstate(void)
  return (value >> 16) & 0xFF;
 }
 
+static int byt_get_turbo_pstate(void)
+{
+ u64 value;
+ rdmsrl(BYT_TURBO_RATIOS, value);
+ return value & 0x3F;
+}
+
 static void byt_set_pstate(struct cpudata *cpudata, int pstate)
 {
  u64 val;
@@ -469,7 +478,7 @@ static struct cpu_defaults byt_params = {
  .funcs = {
   .get_max = byt_get_max_pstate,
   .get_min = byt_get_min_pstate,
-  .get_turbo = byt_get_max_pstate,
+  .get_turbo = byt_get_turbo_pstate,
   .set = byt_set_pstate,
   .get_vid = byt_get_vid,
  },
-- 
1.7.1